File: OpenFlowiOS/ModelBundle.swift
//
//  ModelBundle.swift
//  OpenFlowMobile
//
//  Loaded model + its companion JSONs, with the inference contract baked in.
//  The translation between raw feature values and the model's scaled tensors
//  lives here so callers never touch normalization directly.
//

import Foundation
import CoreML

final class ModelBundle {
    let manifest: MLContract.Manifest
    let trainingConfig: MLContract.TrainingConfig
    let scalers: MLContract.Scalers
    let stationIndex: MLContract.IndexMap
    let basinIndex: MLContract.IndexMap

    var version: String { manifest.modelVersion }

    private let model: MLModel

    /// Initialize from a directory containing the model bundle:
    ///   manifest.json, training_config.json, scalers.json,
    ///   station_index.json, basin_index.json, lstm_model.mlpackage/
    init(directory: URL) throws {
        let dec = JSONDecoder()
        self.manifest = try dec.decode(MLContract.Manifest.self,
            from: try Data(contentsOf: directory.appendingPathComponent("manifest.json")))
        self.trainingConfig = try dec.decode(MLContract.TrainingConfig.self,
            from: try Data(contentsOf: directory.appendingPathComponent("training_config.json")))
        self.scalers = try dec.decode(MLContract.Scalers.self,
            from: try Data(contentsOf: directory.appendingPathComponent("scalers.json")))
        self.stationIndex = try dec.decode(MLContract.IndexMap.self,
            from: try Data(contentsOf: directory.appendingPathComponent("station_index.json")))
        self.basinIndex = try dec.decode(MLContract.IndexMap.self,
            from: try Data(contentsOf: directory.appendingPathComponent("basin_index.json")))

        let mlpackageURL = directory.appendingPathComponent("lstm_model.mlpackage")
        let compiledURL = try MLModel.compileModel(at: mlpackageURL)
        self.model = try MLModel(contentsOf: compiledURL)
    }

    /// Run inference on an assembled window. Returns 14 days of decoded cfs forecasts
    /// keyed by date starting at `forecastStart` (typically tomorrow).
    func forecast(_ window: FeatureWindow, forecastStart: Date) throws -> [DailyForecast] {
        let provider = try buildInputs(window)
        let prediction = try model.prediction(from: provider)
        return try decodeOutput(prediction, forecastStart: forecastStart)
    }

    private func buildInputs(_ window: FeatureWindow) throws -> MLFeatureProvider {
        let schema = manifest.schema

        guard window.encoderRows.count == schema.encoderDays,
              window.encoderRows.allSatisfy({ $0.count == schema.encoderFeatures.count }) else {
            throw InferenceError.shapeMismatch(
                "encoder_input expected [\(schema.encoderDays), \(schema.encoderFeatures.count)]")
        }
        guard window.decoderRows.count == schema.decoderDays,
              window.decoderRows.allSatisfy({ $0.count == schema.decoderFeatures.count }) else {
            throw InferenceError.shapeMismatch(
                "decoder_input expected [\(schema.decoderDays), \(schema.decoderFeatures.count)]")
        }
        guard window.persistenceTargets.count == schema.targetFeatures.count else {
            throw InferenceError.shapeMismatch(
                "persistence_input expected \(schema.targetFeatures.count) values")
        }

        let encoderArr = try MLMultiArray(
            shape: [1, schema.encoderDays as NSNumber, schema.encoderFeatures.count as NSNumber],
            dataType: .float32)
        for d in 0..<schema.encoderDays {
            for c in 0..<schema.encoderFeatures.count {
                let scaled = scale(window.encoderRows[d][c], column: schema.encoderFeatures[c])
                encoderArr[[0, d as NSNumber, c as NSNumber]] = NSNumber(value: Float(scaled))
            }
        }

        let decoderArr = try MLMultiArray(
            shape: [1, schema.decoderDays as NSNumber, schema.decoderFeatures.count as NSNumber],
            dataType: .float32)
        for d in 0..<schema.decoderDays {
            for c in 0..<schema.decoderFeatures.count {
                let scaled = scale(window.decoderRows[d][c], column: schema.decoderFeatures[c])
                decoderArr[[0, d as NSNumber, c as NSNumber]] = NSNumber(value: Float(scaled))
            }
        }

        let persArr = try MLMultiArray(
            shape: [1, schema.targetFeatures.count as NSNumber],
            dataType: .float32)
        for c in 0..<schema.targetFeatures.count {
            let scaled = scale(window.persistenceTargets[c], column: schema.targetFeatures[c])
            persArr[[0, c as NSNumber]] = NSNumber(value: Float(scaled))
        }

        let stationArr = try MLMultiArray(shape: [1], dataType: .int32)
        stationArr[0] = NSNumber(value: stationIndex.index(for: window.siteId))

        let basinArr = try MLMultiArray(shape: [1], dataType: .int32)
        basinArr[0] = NSNumber(value: basinIndex.index(for: window.basinId))

        return try MLDictionaryFeatureProvider(dictionary: [
            "encoder_input": encoderArr,
            "decoder_input": decoderArr,
            "persistence_input": persArr,
            "station_input": stationArr,
            "basin_input": basinArr,
        ])
    }

    private func decodeOutput(_ prediction: MLFeatureProvider, forecastStart: Date) throws -> [DailyForecast] {
        guard let outputValue = prediction.featureValue(for: "persistence_plus_delta"),
              let arr = outputValue.multiArrayValue else {
            throw InferenceError.missingOutput
        }
        let schema = manifest.schema
        guard let minIdx = schema.targetFeatures.firstIndex(of: "Min Flow"),
              let maxIdx = schema.targetFeatures.firstIndex(of: "Max Flow") else {
            throw InferenceError.unexpectedTargets(schema.targetFeatures)
        }

        let cal = Calendar(identifier: .gregorian)
        var forecasts: [DailyForecast] = []
        forecasts.reserveCapacity(schema.decoderDays)
        for d in 0..<schema.decoderDays {
            let zMin = arr[[0, d as NSNumber, minIdx as NSNumber]].doubleValue
            let zMax = arr[[0, d as NSNumber, maxIdx as NSNumber]].doubleValue
            let date = cal.date(byAdding: .day, value: d, to: forecastStart) ?? forecastStart
            forecasts.append(DailyForecast(
                date: date,
                minFlowCFS: invert(zMin, column: "Min Flow"),
                maxFlowCFS: invert(zMax, column: "Max Flow")))
        }
        return forecasts
    }

    /// Apply per-column scaling per the contract. Columns missing from
    /// scalers.json (indicator columns) pass through unchanged.
    private func scale(_ raw: Double, column: String) -> Double {
        guard let params = scalers.columns[column] else { return raw }
        let x: Double
        switch params.transform {
        case .log1p: x = log1p(raw)
        case .identity: x = raw
        }
        return (x - params.mean) / params.scale
    }

    /// Inverse of scale(). Recovers cfs from the model's scaled output.
    private func invert(_ z: Double, column: String) -> Double {
        guard let params = scalers.columns[column] else { return z }
        let x = z * params.scale + params.mean
        switch params.transform {
        case .log1p: return expm1(x)
        case .identity: return x
        }
    }
}

enum InferenceError: LocalizedError {
    case shapeMismatch(String)
    case missingOutput
    case unexpectedTargets([String])

    var errorDescription: String? {
        switch self {
        case .shapeMismatch(let m): return "Input shape mismatch: \(m)"
        case .missingOutput: return "Model produced no 'persistence_plus_delta' output"
        case .unexpectedTargets(let t): return "Unexpected target_features: \(t)"
        }
    }
}

File: OpenFlowAndroid/app/src/main/java/com/tmart234/openflowmobile/ml/ModelBundle.kt
package com.tmart234.openflowmobile.ml

import kotlinx.serialization.json.Json
import org.tensorflow.lite.Interpreter
import java.io.Closeable
import java.io.File
import java.nio.channels.FileChannel
import java.nio.file.StandardOpenOption
import kotlin.math.expm1
import kotlin.math.ln1p

/**
 * Loaded TFLite model + companion JSONs, with the inference contract baked in.
 * The translation between raw feature values and the model's scaled tensors
 * lives here so callers never touch normalization directly.
 */
class ModelBundle private constructor(
    val manifest: Manifest,
    val trainingConfig: TrainingConfig,
    val scalers: Scalers,
    val stationIndex: IndexMap,
    val basinIndex: IndexMap,
    private val interpreter: Interpreter,
) : Closeable {

    val version: String get() = manifest.modelVersion

    /**
     * Run inference on an assembled window. Returns [Manifest.Schema.decoderDays]
     * days of decoded cfs forecasts starting at [forecastStartEpochDay].
     */
    fun forecast(window: FeatureWindow, forecastStartEpochDay: Long): List<DailyForecast> {
        val inputs = buildInputs(window)
        val schema = manifest.schema
        val output = Array(1) { Array(schema.decoderDays) { FloatArray(schema.targetFeatures.size) } }
        val outputs = mapOf("persistence_plus_delta" to output)
        val runner = interpreter.getSignatureRunner(
            interpreter.signatureKeys.firstOrNull() ?: "serving_default"
        )
        runner.run(inputs, outputs)
        return decodeOutput(output, forecastStartEpochDay)
    }

    private fun buildInputs(window: FeatureWindow): Map<String, Any> {
        val schema = manifest.schema

        require(window.encoderRows.size == schema.encoderDays) {
            "encoder_input expected ${schema.encoderDays} rows, got ${window.encoderRows.size}"
        }
        require(window.encoderRows.all { it.size == schema.encoderFeatures.size }) {
            "encoder_input expected ${schema.encoderFeatures.size} cols"
        }
        require(window.decoderRows.size == schema.decoderDays) {
            "decoder_input expected ${schema.decoderDays} rows, got ${window.decoderRows.size}"
        }
        require(window.decoderRows.all { it.size == schema.decoderFeatures.size }) {
            "decoder_input expected ${schema.decoderFeatures.size} cols"
        }
        require(window.persistenceTargets.size == schema.targetFeatures.size) {
            "persistence_input expected ${schema.targetFeatures.size} values"
        }

        val encoderArr = Array(1) {
            Array(schema.encoderDays) { d ->
                FloatArray(schema.encoderFeatures.size) { c ->
                    scale(window.encoderRows[d][c], schema.encoderFeatures[c]).toFloat()
                }
            }
        }
        val decoderArr = Array(1) {
            Array(schema.decoderDays) { d ->
                FloatArray(schema.decoderFeatures.size) { c ->
                    scale(window.decoderRows[d][c], schema.decoderFeatures[c]).toFloat()
                }
            }
        }
        val persArr = Array(1) {
            FloatArray(schema.targetFeatures.size) { c ->
                scale(window.persistenceTargets[c], schema.targetFeatures[c]).toFloat()
            }
        }
        val stationArr = IntArray(1) { stationIndex.indexFor(window.siteId) }
        val basinArr = IntArray(1) { basinIndex.indexFor(window.basinId) }

        return mapOf(
            "encoder_input" to encoderArr,
            "decoder_input" to decoderArr,
            "persistence_input" to persArr,
            "station_input" to stationArr,
            "basin_input" to basinArr,
        )
    }

    private fun decodeOutput(
        output: Array<Array<FloatArray>>,
        forecastStartEpochDay: Long,
    ): List<DailyForecast> {
        val schema = manifest.schema
        val minIdx = schema.targetFeatures.indexOf("Min Flow")
        val maxIdx = schema.targetFeatures.indexOf("Max Flow")
        check(minIdx >= 0 && maxIdx >= 0) {
            "Unexpected target_features: ${schema.targetFeatures}"
        }
        return List(schema.decoderDays) { d ->
            DailyForecast(
                epochDay = forecastStartEpochDay + d,
                minFlowCFS = invert(output[0][d][minIdx].toDouble(), "Min Flow"),
                maxFlowCFS = invert(output[0][d][maxIdx].toDouble(), "Max Flow"),
            )
        }
    }

    /**
     * Apply per-column scaling per the contract. Columns missing from
     * scalers.json (indicator columns) pass through unchanged.
     */
    private fun scale(raw: Double, column: String): Double {
        val params = scalers[column] ?: return raw
        val x = if (params.transform == "log1p") ln1p(raw) else raw
        return (x - params.mean) / params.scale
    }

    /** Inverse of [scale]. Recovers cfs from the model's scaled output. */
    private fun invert(z: Double, column: String): Double {
        val params = scalers[column] ?: return z
        val x = z * params.scale + params.mean
        return if (params.transform == "log1p") expm1(x) else x
    }

    override fun close() {
        interpreter.close()
    }

    companion object {
        private val json = Json { ignoreUnknownKeys = true }

        /**
         * Load from a directory containing the model bundle:
         *   manifest.json, training_config.json, scalers.json,
         *   station_index.json, basin_index.json, lstm_model.tflite
         */
        fun load(directory: File): ModelBundle {
            val manifest = json.decodeFromString(
                Manifest.serializer(),
                File(directory, "manifest.json").readText()
            )
            val trainingConfig = json.decodeFromString(
                TrainingConfig.serializer(),
                File(directory, "training_config.json").readText()
            )
            val scalers: Scalers = json.decodeFromString(
                File(directory, "scalers.json").readText()
            )
            val stationIndex: IndexMap = json.decodeFromString(
                File(directory, "station_index.json").readText()
            )
            val basinIndex: IndexMap = json.decodeFromString(
                File(directory, "basin_index.json").readText()
            )

            val modelFile = File(directory, "lstm_model.tflite")
            val mappedBuffer = FileChannel.open(modelFile.toPath(), StandardOpenOption.READ).use { ch ->
                ch.map(FileChannel.MapMode.READ_ONLY, 0, ch.size())
            }
            // Flex delegate auto-registers if tensorflow-lite-select-tf-ops is on
            // the classpath, which handles manifest.tflite_mode == "select_tf_ops".
            val interpreter = Interpreter(mappedBuffer)

            return ModelBundle(
                manifest = manifest,
                trainingConfig = trainingConfig,
                scalers = scalers,
                stationIndex = stationIndex,
                basinIndex = basinIndex,
                interpreter = interpreter,
            )
        }
    }
}
